google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/SchemaCompact.java

```diff
@@ -76,24 +102,473 @@ private TableId getTableId(String tableName) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  /**
+   * @param field
+   * @return True if fieldtype is supported by BQ Schema
+   */
+  public static boolean isSupportedType(Descriptors.FieldDescriptor field) {
+    Descriptors.FieldDescriptor.Type fieldType = field.getType();
+    if (!SupportedTypes.contains(fieldType)) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Method that checks for proper nesting (no recursive protos) and supported types.
+   *
+   * @param message
+   * @param level Keeps track of current level of nesting.
+   * @param allMessageTypes Keeps track of all message types seen, and make sure they don't repeat
+   *     as recursive protos are not supported.
+   * @return True if fieldtype is supported by BQ Schema
+   * @throws IllegalArgumentException if message is invalid
+   */
+  private static boolean isNestedMessageAccepted(
+      Descriptors.Descriptor message,
+      HashSet<Descriptors.Descriptor> allMessageTypes,
+      String protoScope)
+      throws IllegalArgumentException {
+
+    if (allMessageTypes.size() > 15) {
+      return false;
+    }
+
+    if (allMessageTypes.contains(message)) {
+      throw new IllegalArgumentException(
+          "Proto schema " + protoScope + " is not supported: is a recursively nested message.");
+    }
+    allMessageTypes.add(message);
+    boolean result = isSupportedImpl(message, allMessageTypes, protoScope);
+    allMessageTypes.remove(message);
+    return result;
+  }
+
+  private static boolean isCompatibleWithBQBool(Descriptors.FieldDescriptor.Type field) {
+    if (field == Descriptors.FieldDescriptor.Type.BOOL
+        || field == Descriptors.FieldDescriptor.Type.INT32
+        || field == Descriptors.FieldDescriptor.Type.INT64
+        || field == Descriptors.FieldDescriptor.Type.UINT32
+        || field == Descriptors.FieldDescriptor.Type.UINT64
+        || field == Descriptors.FieldDescriptor.Type.FIXED32
+        || field == Descriptors.FieldDescriptor.Type.FIXED64
+        || field == Descriptors.FieldDescriptor.Type.SFIXED32
+        || field == Descriptors.FieldDescriptor.Type.SFIXED64) {
+      return true;
+    }
+    return false;
+  }
+
+  private static boolean isCompatibleWithBQBytes(Descriptors.FieldDescriptor.Type field) {
+    if (field == Descriptors.FieldDescriptor.Type.BYTES) {
+      return true;
+    }
+    return false;
+  }
+
+  private static boolean isCompatibleWithBQDate(Descriptors.FieldDescriptor.Type field) {
+    if (field == Descriptors.FieldDescriptor.Type.INT32
+        || field == Descriptors.FieldDescriptor.Type.INT64
+        || field == Descriptors.FieldDescriptor.Type.SFIXED32
+        || field == Descriptors.FieldDescriptor.Type.SFIXED64) {
+
+      return true;
+    }
+    return false;
+  }
+
+  private static boolean isCompatibleWithBQDatetime(Descriptors.FieldDescriptor.Type field) {
```

Comment:
also check format:

if (cpp_type == proto2::FieldDescriptor::CPPTYPE_INT64 &&
          (field_format == googlesql::FieldFormat::DATETIME_MICROS ||
           field_format == googlesql::FieldFormat::DEFAULT_FORMAT)) {

Reply:
Discussed with Yiru and we decided we are going to skip format checking for now!